src/service/lightbox-manager-discovery.js

```diff
@@ -0,0 +1,152 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {extensionsFor} from '../extensions';
+import {elementByTag} from '../dom';
+import {isExperimentOn} from '../experiments';
+import {dev} from '../log';
+
+
+const ELIGIBLE_TAGS = [
+  'amp-img',
+  'amp-anim',
+  'amp-ad',
+  'amp-dailymotion',
+  'amp-jwplayer',
+  'amp-kaltura-player',
+  'amp-o2-player',
+  'amp-pinterest',
+  'amp-reach-player',
+  'amp-vimeo',
+  'amp-vine',
+  'amp-youtube',
+  'amp-video',
+  'amp-twitter',
+  'amp-facebook',
+  'amp-instagram',
+];
+
+const ELIGIBLE_TAP_TAGS = {
+  'amp-img': true,
+  'amp-anim': true,
+};
+
+const DEFAULT_VIEWER_ID = 'amp-lightbox-viewer';
+const VIEWER_TAG = 'amp-lightbox-viewer';
+
+/**
+ * Finds elements in the document that meet our heuristics for automatically
+ * becoming lightboxable and adds `lightbox` attribute to them.
+ * It may also install a tap handler on elements that meet our heuristics
+ * to automatically open in lightbox on tap.
+ * @param {!./ampdoc-impl.AmpDoc} ampdoc
+ * @return {!Promise}
+ */
+export function autoDiscoverLightboxables(ampdoc) {
+
+  // Extra safety check, manager should not call this if experiments are off
+  dev().assert(isExperimentOn(ampdoc.win, 'amp-lightbox-viewer'));
+  dev().assert(isExperimentOn(ampdoc.win, 'amp-lightbox-viewer-auto'));
+
+  return new Promise((resolve, unused) => {
+    const viewerId = maybeInstallLightboxViewer(ampdoc);
+
+    const tagsQuery = ELIGIBLE_TAGS.join(',');
+    const matches = ampdoc.getRootNode().querySelectorAll(tagsQuery);
+    for (let i = 0; i < matches.length; i++) {
+      const elem = matches[i];
+      if (!meetsHeuristics(elem)) {
+        continue;
+      }
+      elem.setAttribute('lightbox', '');
+      if (meetsHeuristicsForTap(elem)) {
+        elem.setAttribute('on', 'tap:' + viewerId + '.open');
```

Comment:
This is best to do via default action. E.g. we can add a tap listener via `Action` service and invoke lightbox if conditions are met. I'm ok to leave it like this for now, but add TODO.